internal/v4/stats_test.go

```diff
@@ -90,44 +90,109 @@ func (s *StatsSuite) TestServerStatsStatus(c *gc.C) {
 func (s *StatsSuite) TestServerStatsUpdate(c *gc.C) {
 	ref := charm.MustParseReference("~charmers/precise/wordpress-23")
 	tests := []struct {
-		path       string
-		status     int
-		body       params.StatsUpdateRequest
-		expectBody map[string]interface{}
-		previous   bool
+		path          string
+		status        int
+		body          params.StatsUpdateRequest
+		expectBody    map[string]interface{}
+		previousMonth bool
 	}{{
 		path:   "stats/update",
 		status: http.StatusOK,
-		body: params.StatsUpdateRequest {
-			Timestamp:      time.Now(),
-			CharmReference: charm.MustParseReference("~charmers/wordpress"),
+		body: params.StatsUpdateRequest{
+			Entries: []params.StatsUpdateEntry{{
+				Timestamp:      time.Now(),
+				CharmReference: charm.MustParseReference("~charmers/wordpress"),
+			}}},
+	}, {
+		path:   "stats/update",
+		status: http.StatusOK,
+		body: params.StatsUpdateRequest{
+			Entries: []params.StatsUpdateEntry{{
+				Timestamp:      time.Now(),
+				CharmReference: ref,
+			}},
 		},
-		expectBody: make(map[string]interface {}),
-	},{
+	}, {
 		path:   "stats/update",
 		status: http.StatusOK,
-		body: params.StatsUpdateRequest {
-			Timestamp:      time.Now(),
-			CharmReference: ref,
+		body: params.StatsUpdateRequest{
+			Entries: []params.StatsUpdateEntry{{
+				Timestamp:      time.Now().AddDate(0, -1, 0),
+				CharmReference: ref,
+			}},
 		},
-		expectBody: make(map[string]interface {}),
-	},{
+		previousMonth: true,
+	}}
+
+	ch := storetesting.Charms.CharmDir("wordpress")
+	rurl := newResolvedURL("~charmers/precise/wordpress-23", 23)
+	err := s.store.AddCharmWithArchive(rurl, ch)
+	c.Assert(err, gc.IsNil)
+	err = s.store.SetPerms(&rurl.URL, "read", params.Everyone, rurl.URL.User)
+	c.Assert(err, gc.IsNil)
+
+	var countsBefore, countsAfter charmstore.AggregatedCounts
+	for i, test := range tests {
+		c.Logf("test %d. %s", i, test.path)
+
+		_, countsBefore, err = s.store.ArchiveDownloadCounts(ref, true)
+		c.Assert(err, gc.IsNil)
+
+		rec := httptesting.DoRequest(c, httptesting.DoRequestParams{
```

Comment:
You can still use AssertJSONCall for this. (that should probably be better documented in the AssertJSONCall doc comment)